kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java

```diff
@@ -18,94 +18,276 @@
-            logSegment.deltas.stream());
+        final List<FileStatus> allFiles = Stream.concat(
+                logSegment.checkpoints.stream(),
+                logSegment.deltas.stream()
+            ).collect(Collectors.toList());
         assertLogFilesBelongToTable(logPath, allFiles);
 
-        this.reverseActionsIterable = new ReverseFilesToActionsIterable(
-            tableHelper,
-            allFiles);
-        this.protocolAndMetadata = new Lazy<>(this::loadTableProtocolAndMetadata);
+        this.addRemoveColumnarBatchIterable = new ReverseFilesToColumnarBatchActionsIterable(
+            tableClient,
+            allFiles,
+            ADD_REMOVE_READ_SCHEMA);
+
+        final CloseableIterable<Tuple2<FileDataReadResult, Boolean>>
+            protocolMetadataColumnarBatchIterable = new ReverseFilesToColumnarBatchActionsIterable(
+                tableClient,
+                allFiles,
+            PROTOCOL_METADATA_READ_SCHEMA);
+
+        this.protocolAndMetadata = new Lazy<>(
+            () -> loadTableProtocolAndMetadata(protocolMetadataColumnarBatchIterable)
+        );
     }
 
-    public Lazy<Tuple2<Protocol, Metadata>> lazyLoadProtocolAndMetadata()
-    {
+    /////////////////
+    // Public APIs //
+    /////////////////
+
+    public Lazy<Tuple2<Protocol, Metadata>> lazyLoadProtocolAndMetadata() {
         return this.protocolAndMetadata;
     }
 
-    public CloseableIterator<AddFile> getAddFiles()
-    {
-        final CloseableIterator<Tuple2<Action, Boolean>> reverseActionsIter =
-            reverseActionsIterable.iterator();
-        return new ReverseActionsToAddFilesIterator(dataPath, reverseActionsIter);
+    /**
+     * Returns an iterator of {@link DataReadResult} with schema {@link #ADD_ONLY_DATA_SCHEMA}
+     * representing all the active AddFiles in the table (passing the DataReadResult's selection
+     * vector means it is active).
+     */
+    public CloseableIterator<DataReadResult> getAddFilesAsColumnarBatches() {
+        final CloseableIterator<Tuple2<FileDataReadResult, Boolean>> addRemoveIter =
+            addRemoveColumnarBatchIterable.iterator();
+
+        return new ReverseAddRemoveToAddColumnarBatchIterator(addRemoveIter);
     }
 
-    private Tuple2<Protocol, Metadata> loadTableProtocolAndMetadata()
-    {
+    /**
+     * TEMPORARY. DELETE THIS. We should only expose AddFiles as part of ColumnarBatches via
+     * getAddFilesAsColumnarBatches. That requires updating downstream consumers of this API.
+     *
+     * TODO: delete this API and update downstream consumers to use getAddFilesAsColumnarBatches
+     *       instead.
+     *
+     * Converts the iterator of ColumnarBatches of AddFiles into an iterator of deserialized AddFile
+     * actions.
+     */
+    public CloseableIterator<AddFile> getAddFiles() {
+        return new CloseableIterator<AddFile>() {
+            private final CloseableIterator<DataReadResult> batchesIter = getAddFilesAsColumnarBatches();
```

Comment:
nit: not directly related to this PR... but this iterator-of-iterators is a repeating pattern that is worth making an utility for. Will reduce a lot of code here.

Reply:
I agree. FYI, this `getAddFiles` method will soon be entirely deleted

Reply:
yeah. but if you see this pattern at other places..please make it an utility.